File: src/engine/blit.c
// see description of this algorithm in http://bitboxconsole.blogspot.fr/2013/10/an-efficient-scanline-blitting-algorithm.html

/* TODO : 
include DXT
include tilesprite
make a font element (avec sub packbits : array de chars = packbits independants, taille H donnee)
*/
#include "blit.h"

#include <stdint.h>
#include <stddef.h> // NULL

// XXX Add some locking within frame ? 


/*

 lists : object array characteristics
 object list : sorted y, slowly inserted / removed, can be big <- as array. 
    optimisation+ tard : keep holes / linked list
 
 active list : allocate and insert semifast. need to remove also. insert : 1/line

 blit : FAST, ~ size active, kept sorted, emptied auto, no remove, insert,. frequent (N times/line) allocate & insert sorted. no re-sorting
    ~16 max ? 

 lifo transp stack : fast, limited?), emptied never popped, pushed seq : an array .

 */

typedef struct {
    // ** long standing variables

    // list of objects blit.
    int nb_objects; // maximum used objects. There may be holes before this however.
    object *objects[MAX_OBJECTS]; // Sorted by Y. Objects shall not be added or modified within display

    // ** frame-lived variables (ie they are reset each frame)

    int current_line; // current line blitting

    // active list is the list of object currently being blit. 
    // an object is not active if the displayed line is before its topmost line, or after the bounding box.
    object *activelist_head; // top of the active list
    int next_to_activate; // next object to put in active list

} Blitter;

// global local variable
static Blitter blt;

void blitter_init()
{   
    // initialize empty objects array
    for (int i=0;i<MAX_OBJECTS;i++) 
    {
        blt.objects[i]=&real_objects[i]; 
        real_objects[i].y=INT16_MAX;
    }
    blt.nb_objects = 0; // first unused is first.
}

// return new object pointer
// append to end of list ; list ends up unsorted now
// DON'T insert something twice !
{
    if (blt.nb_objects<MAX_OBJECTS)
        return blt.objects[blt.nb_objects++]; // index of free object IN !
    else 
        return 0;
}

void blitter_remove(object *o)
{
    // ! Should do that between frames (only frame-based variables will be updated)
    o->y = INT16_MAX;   
}

// http://en.wikipedia.org/wiki/Insertion_sort
// insertion sort objects array by Y. Simplest form, just sorting
void blitter_sort_objects_y()
{
    // consider empty values as bigger than anything (Y = INT16_MAX)

    object *valueToInsert; 
    int holePos;
    int real_nbobjects = 0; // count the real number of objects

    // The values in blt.objects[i] are checked in-order, starting at the second one
    for (int i=1;i<blt.nb_objects;i++) 
    {
        // at the start of the iteration, objects[0..i-1] are in sorted order
        // this iteration will insert blt.objects[i] into that sorted order
        // save blt.objects[i], the value that will be inserted into the array on this iteration
        valueToInsert = blt.objects[i];

        // now mark position i as the hole; blt.objects[i]=blt.objects[holePos] is now considered empty
        holePos=i;

        // keep moving the hole down until the valueToInsert is larger than 
        // what's just below the hole or the hole has reached the beginning of the array
        // null pointers are considered larger than anything (except null pointers but we've made the test already)
        while (holePos > 0 && (!blt.objects[holePos - 1] || valueToInsert->y < blt.objects[holePos - 1]->y))
        { //value to insert doesn't belong where the hole currently is, so shift 
            blt.objects[holePos] = blt.objects[holePos - 1]; //shift the larger value up
            holePos -= 1;       //move the hole position down
        }

        // hole is in the right position, so put valueToInsert into the hole
        blt.objects[holePos] = valueToInsert;
        
        real_nbobjects+=1; // count number of real objects (skipped the holes)

        // blt.objects[0..i] are now in sorted order
    }
    blt.nb_objects=real_nbobjects;
}



void blitter_frame()
{
    // ensure objectlist is sorted by y 
    blitter_sort_objects_y();

    blt.current_line = 0;
    // reset activelist, next to blit
    blt.activelist_head = (object*)0;
    blt.next_to_activate = 0;

    // rewind all objects. nb objects is up to date (no holes, since we just sorted them)
    object *o;
    for (int i=0;i<blt.nb_objects;i++)
    {
        o=blt.objects[i];
        o->frame(o,o->y<0?-o->y:0); // first line is -y if negative
    }
}


void activelist_add(object *o)
// insert sorted by Z
{
    object **head = &blt.activelist_head;
    // find insertion point
    while (*head && (*head)->z < o->z) 
        head = &(*head)->activelist_next;

    // insert effectively
    o->activelist_next = *head;
    *head = o;
} 


void blitter_line()
{
    // drop past objects from active list. not sorted by Y
    object *prev=NULL;
    for (object *o=blt.activelist_head;o;o=o->activelist_next)
    {
        if (blt.current_line >= o->y+o->h)
        {           
            // remove this object from active list
            if (o==blt.activelist_head)
            {   // change head ?
                blt.activelist_head = o->activelist_next;
            } else {
                prev->activelist_next=o->activelist_next;
            }
        }
        prev=o;
    }

    
    // add new active objects
    while (blt.next_to_activate<blt.nb_objects && blt.current_line>=blt.objects[blt.next_to_activate]->y)
    {
        activelist_add(blt.objects[blt.next_to_activate]);
        //printf("activate %d\n",blt.objects[blt.next_to_activate]->x);
        blt.next_to_activate++;
    }


    // now trigger each activelist, in Z descending order
    for (object *o=blt.activelist_head;o;o=o->activelist_next)
    {
        o->line(o); 
    }

    // next line
    blt.current_line++;
}


// --- misc implementations & tests 

void dummy_frame(object *o,int first_line) {}

void opaquerect_line (object *o)
{
    pre_draw(o, o->x,o->x+o->w,OPAQUE);
}

void color_blit(object *o, int16_t x1, int16_t x2) 
{
    // memset of draw_buffer to color o->b
    //printf("BLIT(%d - %d)=%x\n",x1,x2,o->a);
    for (int i=x1;i<=x2;i++)
        draw_buffer[i]=o->a; // 32->16
}

// fixme colors from o
void checkers_blit(object *o, int16_t x1, int16_t x2) // unoptimized !!
{
    // unroll loop
    for (int i=x1;i<=x2;i++)
        draw_buffer[i] = ((blt.current_line/8)+(i/8))%2?(o->a&0xffff):(o->a>>16);  
}

int new_opaquerect(object *o, int16_t x, int16_t y, int16_t w, int16_t h,int16_t z, uint16_t color)
{
    int i = blitter_insert((object *)o);    
    if (i<0) return i;

    // generic attributes
    o->x=x; o->y=y;
    o->h=h; o->w=w;
    o->z=z;

    // special attributes
    o->a = (int)color;
    
    o->frame=dummy_frame;
    o->line=opaquerect_line;
    o->blit=color_blit;
    return i;
}

int new_checkerrect(object *o, int16_t x, int16_t y, int16_t w, int16_t h,int16_t z, uint16_t color1, uint16_t color2)
{
    int i = blitter_insert((object *)o);    
    if (i<0) return i;

    // generic attributes
    o->x=x; o->y=y;
    o->h=h; o->w=w;
    o->z=z;

    // special attributes
    o->a = color1<<16 | color2;
    
    o->frame=dummy_frame;
    o->line=opaquerect_line;
    o->blit=checkers_blit;
    return i;
}


// -------------------------------------------
/* mask data : to be used with a fill function. uses a & b variables of objects 
 *  mask is a uint16_t separate array, pointed by a. b is an internal state variable.
 *  bitmap data can be implmented differently : raw, DXT data, ...
 *  easier to combine if blit can be defined independently, ie can determine source data uniquely fom x
 * 
 *  list of skip/blit alternates, with eol bit. start with skip
 *  mask_record : u16 : skip:7, eol:1, data:8  
 *  fin du masque determine par y/nb lines
 */
#define MASK_EOL (1<<7)
#define MASK_LEN(o)  (o->b & 0x7f)
#define MASK_SKIP(o) (o->b >> 9)

#define MASK_PTR(o) (o->b) // current pointer to mask data

inline void mask_frame(object *o, int start_line)
{
    uint16_t *p=(uint16_t *)o->a; // reset pointer to frame start.
    for (;start_line;start_line--) // scan to line. shall not be finished before. skips N eols
    {
        for (
            ;
            !(*p & MASK_EOL);
            p++
            ); // skip a line. gets to the next blit & check eol
    }
    o->b = (uintptr_t)p;
}

inline void mask_line(object *o)
{
    int x=o->x; // copy to the next 
    do {
        x += MASK_SKIP(o);
        pre_draw (o, o->x,o->x+MASK_LEN(o),OPAQUE); // will blit from ptr, maybe cut & restricted, maybe several times
        x += MASK_LEN(o);
        MASK_PTR(o) +=  sizeof(uint16_t); 
        // skip a line. gets to the next blit & check eol
    } while (*(uint16_t*)MASK_PTR(o) & MASK_EOL); // eol
}


// -- masked rle
// combines a packbit with a mask data
// skip lines ok, blit lines ok, combined with data
// advance data along 



// ---------------------------------------------------------------------------

File: src/engine/blit.py
#encoding: utf-8
'''sample blit algorithms'''

MAXINT = 32768
MININT = -32767
SCREEN_WIDTH = 640

class Blitter : 
	'a blitter is responsible for calling each object'
	'this is a non transparent blitter, ie no transparent is kept.'
	def __init__(self) : 
		# long standing variables
		self.objectlist = [] # sorted by y

		# frame-lived variables
		self.activelist = [] # currently blitting object, sorted by z => wat about opaque 
		self.current_line = 0 # current line to blit. (use a global ?)
		self.next_to_blit = 0

		# line-lived variables
		self.opaque = [] # list of x1,x2 opaque blits on the line. the list is non overlapping and has growing x
		self.transparent_fifo = [] # (x1,x2,b) fifo, filled top to bottom
		
	def insert_object(self,o) : 
		# insert sorted, should use insertion sort (only add during frame !)
		self.objectlist.append(o)

	def frame(self) : 
		# ensure sort by y (should be already sorted)
		self.objectlist.sort(key=lambda o:o.y)

		# rewind all drawable frames
		for o in self.objectlist : o.frame(0)
		self.activelist = [] 
		self.next_to_blit = 0 # next object to blit in objectlist
		self.current_line = 0

	def line(self) : 
		# drop past objects from active list
		for i,o in enumerate(self.activelist) : 
			if self.current_line > o.y+o.h : 
				del self.activelist[i] # remove from list

		# add new active objects
		while (self.next_to_blit<len(self.objectlist) and self.current_line>=self.objectlist[self.next_to_blit].y) :
			self.activelist.append(self.objectlist[self.next_to_blit])
			self.next_to_blit += 1
			self.activelist.sort(key=lambda o:o.z)

		# will gather opaque blits & order blits
		
		# reset opaque list to screen coordinates
		self.opaque = [(MININT,0), (SCREEN_WIDTH,MAXINT)] 

		for o in self.activelist : 
			o.line(self) # will call pre_draw and blit opaques

		while self.transparent_fifo : # in reality, can just scan in reverse.
			x1,x2,b = self.transparent_fifo.pop() 
			b.blit(x1,x2) 


	def pre_draw (self,x1,x2,b,is_opaque) : 
		'declare opaque for this line & call blitting'
		'args : x1,x2,is_opaque  : declares span. '
		'b : reference for callback to blit'
		# there will always be a blit after x2, since screen boundaries is an infinite blit
		# transparent = push a masked line to the draw FIFO, which will be emptied after the make opaque
		print "\npredraw %d-%d"%(x1,x2),"opaque" if is_opaque else "transp"
		# skip all past indices in the list. FIXME keep index in mem 
		bltindex = 0
		while (x1>=self.opaque[bltindex][1]) : bltindex +=1


		while True : 
			print self.opaque, 'idx:',bltindex,'range', x1, x2
			# are we starting within a blit or before ? 
			if (self.opaque[bltindex][0]<=x1) : 
				# skip to the blit end
				x1 = self.opaque[bltindex][1]
				# advance blit
				bltindex += 1

			# ok now we passed, we're sure to be after a blit and not inside it. 
			# we're also sure to have a preceding one

			# blitte jusqu'au prochain obstacle ou la fin 

			if x1>=x2 : break # already finished (includes end of line)

			# there is a next blit here since x1 would have been MAXINT if it was the last
			if x2<=self.opaque[bltindex][0] : # do we finish before or at the start of the next opaque 
				print 'we finish now'
				# blit all
				# blit start is end of preceding blit ?
				if is_opaque : 
					b.blit(x1,x2)
					# mark opaque
					if (x1==self.opaque[bltindex-1][1]) : 
						# extend the current blit
						self.opaque[bltindex-1]=(self.opaque[bltindex-1][0],x2)
					else : # create a new one up to end
						self.opaque.insert(bltindex,(x1,x2))
				else : 
					print 'pushed f',x1,x2
					self.transparent_fifo.append((x1,x2,b))
				break
			else : 
				if is_opaque : 
					# restrict & extend next one (we ARE in line with preceding here, so we extend)
					b.blit(x1,self.opaque[bltindex][0])
					# mark opaque
					self.opaque[bltindex]=(x1, self.opaque[bltindex][1])
				else : 
					print 'pushed',x1,  self.opaque[bltindex][0]
					self.transparent_fifo.append((x1,  self.opaque[bltindex][0], b))
				x1 = self.opaque[bltindex][1]
				bltindex += 1
		print 'now',self.opaque


class Drawable : 
	def __init__(self) : 
		self.x=self.y=self.z=0
		self.w=self.h=0

	def line(self,blitter) : 
		'calls make_opaque of the blitter (& blit it ?)'
		pass

	def blit(self, x1, x2) : 
		'''really blit it, transparently or not, for current line.
		Always called with growing x, without overlap. 
		e.g. : if we first blit from x1 to x2, then the next x'1,x'2 blit will be so that x'1>=x2 (also x2>x1)
		'''
		pass

	def frame(self, line) : 
		'''do whatever is needed to rewind for a new frame. 
		Also fast forwards to line "line", without blitting
		 - if we start blitting at line y by example.'''
		pass

class TestDrawable (Drawable) :
	"color square, aligned on axes"
	def __init__(self, color) : 
		Drawable.__init__(self)
		self.color = color

	def line(self, blitter) :
		blitter.pre_draw(self.x, self.x+self.w, self, True)

	def blit(self, x1, x2) : 
		if x1<x2 :
			print '** blit of %s from'%self.color,x1, 'to', x2 


class TransparentDrawable (Drawable) :
	"Transparent square aligned on axes"
	def __init__(self, color) : 
		Drawable.__init__(self)
		self.color = color

	def line(self, blitter) :
		blitter.pre_draw(self.x, self.x+self.w, self, False)

	def blit(self, x1, x2) : 
		if x1<x2:
			print '++ transparent blit of %s from'%self.color,x1, 'to', x2 

b=Blitter()

blits = [(0,10),(5,15),(25,30),(24,26),(12,32),(8,35),(0,640)] # x1,x2 sorted by low to high, same y
for z,(x1, x2) in enumerate(blits) : 
	o = TestDrawable('black')
	o.x = x1
	o.w = x2-x1
	o.z = 2*z
	o.h=10
	b.insert_object(o)

o=TransparentDrawable('grey')
o.x = -50 ; o.w = 400 ; o.z = 9 ; o.h = 10
b.insert_object(o)

o=TransparentDrawable('blue')
o.x = -50 ; o.w = 640+50 ; o.z = 11; o.h = 10
b.insert_object(o)


b.frame()
b.line()



